src/core/connection_pool.c

```diff
@@ -0,0 +1,350 @@
+/*++
+
+    Copyright (c) Microsoft Corporation.
+    Licensed under the MIT License.
+
+Abstract:
+
+    Definitions for the MsQuic Connection Pool API, which allows clients to
+    create a pool of connections that are spread across RSS cores.
+
+--*/
+
+#include "precomp.h"
+
+#ifdef QUIC_CLOG
+#include "connection_pool.c.clog.h"
+#endif
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+QUIC_API
+MsQuicConnectionPoolCreate(
+    _In_ HQUIC Registration,
+    _In_ QUIC_CONNECTION_CALLBACK_HANDLER Handler,
+    _In_opt_ void* Context,
+    _In_opt_ const char* ServerName,
+    _In_opt_ const QUIC_ADDR* ServerAddress,
+    _In_ uint16_t ServerPort,
+    _In_ uint16_t NumberOfConnections,
+    _Out_writes_bytes_(NumberOfConnections * sizeof(HQUIC))
+        HQUIC** ConnectionPool
+    )
+{
```

Comment:
IMO, this is a very big function (>300 lines). It could easily be split into multiple helpers to be easier to understand and less error prone (keep resource management and error handling more scoped, limit the number of imbricated scopes):

At least these would be easy to split:
- Getting the RSS config
- Testing a port

Reply:
Let's move the logic to get a local address and interface index into its own function.

Reply:
I thought we avoided single-use helper functions, but I don't feel strongly about it.

Reply:
I don't know where this "avoid single-use helper" rule come from ( as in I have no context about what it tries to help with, I'm interested if you know more). In my experience:
- avoiding to add single use helpers *in a "util / common" file is valuable to avoid cluttering. If it is used only in one place, keep it local
- as for everything, there is a balance. It is just as bad having to jump between 50 different 3 lines functions than reading 500 lines long function.
- but breaking down a long function in a succession of sub-tasks, each implemented by a helper function, clarifies what the function does (helper function names are self-documenting), enforce some abstraction between independent tasks (ressource management, error handling, variable being used...)

Reply:
I am a proponent for breaking this down into more easily testable and understandable functions.

Reply:
Well I've split out some of the functions. I'll go over it one more time and see if I can split out anything else that's overly complex.